File: bench/langs/ledger.clj
;; Keeps the running balance for one account.
(ns bench.ledger
  (:require [clojure.test :refer [deftest is]]))

(def limit 5000)

(defmulti describe :kind)

(defmethod describe :entry [entry]
  (format-entry entry))

(defn- refuse [owner amount]
  (warn owner)
  0)

(defn- commit-entry [owner amount]
  (write-entry owner amount)
  amount)

;; Bills the account and returns what is left.
(defn charge [owner amount]
  (if (> amount limit)
    (refuse owner amount)
    (commit-entry owner amount)))

(deftest charge-refuses-over-limit
  (is (= 0 (charge "a" 9000))))

File: bench/langs/Ledger.cs
namespace Bench;

/// <summary>Keeps the running balance for one account.</summary>
public class Ledger
{
    const int Limit = 5000;

    private readonly string owner;

    public Ledger(string owner)
    {
        this.owner = owner;
    }

    /// <summary>Bills the account and returns what is left.</summary>
    public int Charge(int amount)
    {
        if (amount > Limit)
        {
            return Refuse(amount);
        }
        return Commit(amount);
    }

    private int Refuse(int amount)
    {
        Warn(owner);
        return 0;
    }

    private int Commit(int amount)
    {
        Write(owner, amount);
        return amount;
    }
}
